accelerometer/processing.py

```diff
@@ -0,0 +1,202 @@
+from scipy.ndimage import median_filter
+from tqdm.auto import tqdm
+tqdm.pandas()
+
+
+GRAVITY_CUTOFF_HZ = 0.5
+NOISE_CUTOFF_HZ = 20
+MAXVAL, MINVAL = 6, -6  # Note: GENEActiv has a range of +/-6g whereas Axivity and Actigraph +/-8 or more
+
+
+class Processing():
+    def __init__(self, npyFile, sampleRate, epochPeriod, epochFile):
+        self.npyFile = npyFile
+        self.sampleRate = sampleRate
+        self.epochPeriod = epochPeriod
+        self.epochFile = epochFile
+
+        data = loadNpyToFrame(npyFile)
+        data = regularizeSampleRate(data, sampleRate)
+        data = filterNoise(data, sampleRate)
+        #TODO
+        # data = calibrateGravity(...)
+        #TODO
+        # data = detectNonwear(...)
+
+        epochFeats = extractFeatures(data, sampleRate, epochPeriod)
+
+        epochFeats.to_csv(epochFile)
+
+
+def loadNpyToFrame(npyFile):
+    data = pd.DataFrame(np.load(npyFile))
+    data['time'] = data['time'].astype('datetime64[ms]')
+    data.set_index('time', inplace=True)
+    return data
+
+
+def regularizeSampleRate(data, sampleRate):
+    samplePeriod = int(1000_000_000/sampleRate)  # in nanos
+    return data.resample(f'{samplePeriod}N').nearest(limit=1)
+
+
+def filterNoise(data, sampleRate):
+    # Clip unrealistically high values
+    data[:] = np.clip(data.values, MINVAL, MAXVAL)
+    # Noise removal by median filtering
+    data[:] = median_filter(data.values, (5,1), mode='nearest')
+    # Noise removal by lowpass filtering
+    mask = data.isna().any(1)
+    data.fillna(method='ffill', inplace=True)  # temporarily fill NaNs for filtering
+    data[:] = butterfilt(data.values, NOISE_CUTOFF_HZ, sampleRate, axis=0)
+    data[mask] = np.nan
+    return data
+
+
+#TODO
+def calibrateGravity():
+    pass
+
+
+#TODO
+def detectNonwear():
+    pass
+
+
+def extractFeatures(data, sampleRate, epochPeriod):
+    epochs = data.groupby(pd.Grouper(freq=f'{epochPeriod}S'))
+
+    def extract(data):
+        xyz = data.values
+
+        return pd.Series({
+            **basicFeatures(xyz, sampleRate),
+            **sanDiegoFeatures(xyz, sampleRate),
+            **unileverFeatures(xyz, sampleRate)
+        })
+
+    epochFeats = epochs.progress_apply(extract)
+
+    return epochFeats
+
+
+def basicFeatures(xyz, sampleRate):
+    feats = {}
+    feats['xMean'], feats['yMean'], feats['zMean'] = np.mean(xyz, axis=0)
+    feats['xStd'], feats['yStd'], feats['zStd'] = np.std(xyz, axis=0)
+    feats['xRange'], feats['yRange'], feats['zRange'] = np.ptp(xyz, axis=0)
+
+    x, y, z = xyz.T
+    feats['xyCorr'] = np.nan_to_num(np.corrcoef(x, y)[0,1])
+    feats['yzCorr'] = np.nan_to_num(np.corrcoef(y, z)[0,1])
+    feats['zxCorr'] = np.nan_to_num(np.corrcoef(z, x)[0,1])
+
+    m = np.linalg.norm(xyz, axis=1)
+    feats['enmoTrunc'] = np.mean(np.maximum(m - 1.0, 0.0))
+    feats['mad'] = stats.median_abs_deviation(m)
+
+    return feats
+
+
+def sanDiegoFeatures(xyz, sampleRate):
```

Comment:
It is great that you have faithfully reimplemented the San Diego and Unilever functions. We can test these for backwards compatibility. Soon after this pull request it would then be good to condense/simplify this feature set, probably in the work @angerhang has planned.